### PR TITLE
feat(lambda-at-edge): use new aws s3 client for faster require time

### PR DESCRIPTION
**Pull request by dphang**
## Description

This closes https://github.com/serverless-nextjs/serverless-next.js/issues/580.

* Uses the client here in AWS SDK JS v3, or `@aws-sdk/client-s3` (https://github.com/aws/aws-sdk-js-v3/tree/master/clients/client-s3) which is a more modularized client. It will allow us to import only the commands needed (GetObject, PutObject) instead of heavy S3 client in `aws-sdk`, and do it dynamically instead of in every function invocation (i.e before the handler). E.g this will only need it for `getStaticPaths` cases.
* Tests in the linked Issue have shown it takes < 3 ms to load all of these. Unfortunately it does increase handler size from ~150 kB to ~500 kB because there are still a lot of other types for other S3 APIs (but the command themselves only import 2) (but code size has very little impact on cold start times compared to require size, i.e requiring `aws-sdk/S3` will load all APIs. In another PR, we could use terser with Rollup to reduce this to ~180 kB (and we can provide both minified and normal handlers).

Note, I had to use `rollup-plugin-typescript2` (a fork of original plugin) as I was getting these errors and could not resolve it:

```
[!] (plugin typescript) Error: Could not load /Users/danielphang/IdeaProjects/serverless-next.js/packages/libs/lambda-at-edge/node_modules/@aws-sdk/client-s3/S3Client.ts (imported by src/default-handler.ts): Debug Failure. False expression: Expected fileName to be present in command line
```

Though this is better in the sense that it prints more informative error messages. I also added `json` rollup plugin as it is needed to bundle the AWS client correctly.

## Tests

Updated the unit tests in `origin-response` handler test.

Currently in progress of testing end-to-end.

### CODE REVIEW

**Comment by dphang** [comment]
This has been tested on this test app:

https://d2fn5f8m5hw83u.cloudfront.net/tags/128128218 (try replacing the end number with a random number, it will load fallback until it gets the data)

File: packages/libs/lambda-at-edge/src/default-handler.ts [hunk 1]
```diff
@@ -19,10 +19,10 @@ import {
   OriginResponseEvent,
   PerfLogger
 } from "../types";
-import S3 from "aws-sdk/clients/s3";
 import { performance } from "perf_hooks";
 import { ServerResponse } from "http";
 import jsonwebtoken from "jsonwebtoken";
+import { Readable } from "stream";
```
Comment (dphang) [inline packages/libs/lambda-at-edge/src/default-handler.ts:25]: This is just a type, should be fine to import here.

File: packages/libs/lambda-at-edge/src/default-handler.ts [hunk 2]
```diff
@@ -422,7 +429,14 @@ const handleOriginResponse = async ({
...
       Key: `static-pages${hasFallback.fallback}`
     };
-    const { Body } = await s3.getObject(s3Params).promise();
+    const { GetObjectCommand } = await import(
+      "@aws-sdk/client-s3/commands/GetObjectCommand"
+    );
+    const { Body } = await s3.send(new GetObjectCommand(s3Params));
+
+    // Body is stream per: https://github.com/aws/aws-sdk-js-v3/issues/1096
+    const getStream = await import("get-stream");
```
Comment (dphang) [inline packages/libs/lambda-at-edge/src/default-handler.ts:438]: I am using get-stream package: https://bundlephobia.com/result?p=get-stream@6.0.0 which is small

File: packages/libs/lambda-at-edge/src/default-handler.ts (hunk 1 above)
Comment (danielcondemarin) [reply packages/libs/lambda-at-edge/src/default-handler.ts:25]:
```suggestion
import type { Readable } from "stream";
```

File: packages/libs/lambda-at-edge/src/default-handler.ts [hunk 3]
```diff
@@ -156,8 +156,10 @@ const router = (
 export const handler = async (
   event: OriginRequestEvent | OriginResponseEvent
 ): Promise<CloudFrontResultResponse | CloudFrontRequest> => {
+  // @ts-ignore
```
Comment (danielcondemarin) [inline packages/libs/lambda-at-edge/src/default-handler.ts:159]: do we need these @ts-ignores ?

Comment (dphang) [reply packages/libs/lambda-at-edge/src/default-handler.ts:159]: Hm, I saw errors when building at one point due to this being from json file, maybe this one is not needed anymore, let me try to remove it

File: packages/libs/lambda-at-edge/src/default-handler.ts (hunk 1 above)
Comment (dphang) [reply packages/libs/lambda-at-edge/src/default-handler.ts:25]: Oh, I thought Rollup would optimize this but I guess explicitly importing just type is good practice, too.

**Review by danielcondemarin — DISMISSED** [review]
Just a few minor things but overall lgtm 👍

**Review by danielcondemarin — APPROVED** [review]

**Comment by dphang** [comment]
Thanks for merging!

**Pull request merged**